### PR TITLE
Allow DelayedLoadWrapper to late-construct its content

### PR DESCRIPTION
**Pull request by peppy**
Not only does this tidy up the `DelayedLoadUnloadWrapper` flow, but it allows us to (in both cases) avoid the overhead of construction of the content drawable until a load actually occurs.

Tested against my beatmap carousel changes, this reduces finalizer disposal to zero (and greatly reduces allocs in fast scrolling) as expected.

- Make sure you are okay with the Relative/AutoSize propgation. I haven't changed anything for the old-style ctor for compatibility reasons, but now the transfer is done at load time, overwriting anything set locally. Maybe this needs to be mentioned somewhere in the xmldoc?
- Should we obsolete the old ctor?

### CODE REVIEW

**Comment by bdach** [comment]
The change itself seems fine to me.

As for the obsoletion of the old ctor - we actually have game-side usages that would arguably become uglier if we forced late construction. See [`LeaderboardScore` for one](https://github.com/ppy/osu/blob/74a0886cb8f41018a95296f24e1b113af83307df/osu.Game/Online/Leaderboards/LeaderboardScore.cs#L117-L133) - the constructed-but-not-loaded drawable is then [used to register an `OnLoadComplete`](https://github.com/ppy/osu/blob/74a0886cb8f41018a95296f24e1b113af83307df/osu.Game/Online/Leaderboards/LeaderboardScore.cs#L230) callback; forcing late construction would probably make this way worse to read (would have to subscribe to `DelayedLoadComplete` and then nested-subscribe to `OnLoadComplete`).

Also, just so there's no blockage - this technically conflicts with #3936 but the resolution is relatively easy. Here's the combined diff:

<details>
<summary>combined diff</summary>

```diff
diff --cc osu.Framework/Graphics/Containers/DelayedLoadUnloadWrapper.cs
index 49d012b81a,f7c72c2e94..d383017ee8
--- a/osu.Framework/Graphics/Containers/DelayedLoadUnloadWrapper.cs
+++ b/osu.Framework/Graphics/Containers/DelayedLoadUnloadWrapper.cs
@@@ -56,8 -58,10 +56,8 @@@ namespace osu.Framework.Graphics.Contai
              }
          }
  
-         private bool contentLoaded;
 -        public override Drawable Content => base.Content ?? (Content = createContentFunction());
 -
          private ScheduledDelegate scheduledLifetimeUpdate;
+         private ScheduledDelegate scheduledUnloadCheckRegistration;
  
          protected override void EndDelayedLoad(Drawable content)
          {
diff --cc osu.Framework/Graphics/Containers/DelayedLoadWrapper.cs
index 7895f5e13c,904f4ec90c..d3922c3e20
--- a/osu.Framework/Graphics/Containers/DelayedLoadWrapper.cs
+++ b/osu.Framework/Graphics/Containers/DelayedLoadWrapper.cs
@@@ -121,14 -83,16 +121,18 @@@ namespace osu.Framework.Graphics.Contai
  
          protected void BeginDelayedLoad()
          {
-             if (loadTask != null) throw new InvalidOperationException("Load is already started!");
+             if (DelayedLoadTriggered || DelayedLoadCompleted)
+                 throw new InvalidOperationException("Load has already started!");
  
 +            Content ??= createFunc();
 +
+             DelayedLoadTriggered = true;
              DelayedLoadStarted?.Invoke(Content);
  
+             cancellationTokenSource = new CancellationTokenSource();
+ 
              // The callback is run on the game's scheduler since DLUW needs to unload when no updates are being received.
-             loadTask = LoadComponentAsync(Content, EndDelayedLoad, scheduler: Game.Scheduler);
+             LoadComponentAsync(Content, EndDelayedLoad, scheduler: Game.Scheduler, cancellation: cancellationTokenSource.Token);
          }
  
          protected virtual void EndDelayedLoad(Drawable content)
```

</details>

(in short, take both additions in DLW and both removals in DLUW).

**Review by bdach — APPROVED** [review]

File: osu.Framework/Graphics/Containers/DelayedLoadWrapper.cs [hunk 1]
```diff
@@ -21,37 +21,75 @@ public class DelayedLoadWrapper : CompositeDrawable
...
+            this.timeBeforeLoad = timeBeforeLoad;
 
             AddLayout(optimisingContainerCache);
             AddLayout(isIntersectingCache);
         }
 
         public override double LifetimeStart
         {
-            get => Content.LifetimeStart;
+            get => Content?.LifetimeStart ?? base.LifetimeStart;
```
Comment (bdach) [inline osu.Framework/Graphics/Containers/DelayedLoadWrapper.cs:61]: I sort of glossed over this before, but this is a bit of an API inconsistency in that setting a DLW's lifetime behaves differently to setting DLUW's lifetime before the content is constructed.

* To set a DLW's content's lifetime before it's constructed you have to make sure to specify the lifetime in the content-constructing delegate, as setting lifetime on DLW before construction will throw on a nullref.
* To set a DLUW's content's lifetime before it's constructed you can just set `Lifetime{Start,End}` on the DLUW itself and the logic therein will ensure it's transferred upon the content being constructed.

Maybe we want to move the lifetime logic, along with the lifetime-setting schedule, to DLW entirely?

Comment (peppy) [reply osu.Framework/Graphics/Containers/DelayedLoadWrapper.cs:61]: Yeah, probably. The whole lifetime propagation is a bit haphazard in general though.

**Review by smoogipoo — APPROVED** [review]
lgtm

**Pull request merged**